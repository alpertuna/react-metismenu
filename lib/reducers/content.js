'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _content = require('../actions/content');

var _flattenContent = require('./flattenContent');

var _flattenContent2 = _interopRequireDefault(_flattenContent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file reducers/content.js
 * @author H.Alper Tuna <halpertuna@gmail.com>
 * Date: 16.12.2016
 */

/* eslint-env browser */

var item = function item(state, action) {
  switch (action.type) {
    case 'CHANGE_SUBMENU_VISIBILITY':
      {
        return Object.assign({}, state, {
          subMenuVisibility: state.id === action.id ? action.subMenuVisibility : action.trace.indexOf(state.id) !== -1
        });
      }
    case 'CHANGE_ACTIVE_LINK_FROM_LOCATION':
    case 'CHANGE_ACTIVE_LINK':
      {
        return Object.assign({}, state, {
          active: state.id === action.id,
          hasActiveChild: action.trace.indexOf(state.id) !== -1
        });
      }
    default:
      {
        return state;
      }
  }
};

var findItem = function findItem(content, value, prop) {
  return content.find(function (i) {
    return i[prop] === value;
  });
};

var content = function content() {
  var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var action = arguments[1];

  switch (action.type) {
    case 'UPDATE_CONTENT':
      {
        return (0, _flattenContent2.default)(action.content);
      }
    case 'CHANGE_SUBMENU_VISIBILITY':
      {
        return state.map(function (i) {
          return item(i, action);
        });
      }
    case 'CHANGE_ACTIVE_LINK_FROM_LOCATION':
    case 'CHANGE_ACTIVE_LINK':
      {
        var activeItem = void 0;
        if (action.type === 'CHANGE_ACTIVE_LINK_FROM_LOCATION') {
          var locationSets = [window.location.pathname + window.location.search, // /path?search
          window.location.hash, // #hash
          window.location.pathname + window.location.search + window.location.hash];
          activeItem = state.find(function (i) {
            return locationSets.indexOf(i.to) !== -1;
          });
        } else {
          activeItem = findItem(state, action.value, action.propName);
        }

        // If metismenu user tries to activate non-exist item
        if (!activeItem) return state;

        var _activeItem = activeItem,
            id = _activeItem.id,
            parentId = _activeItem.parentId,
            trace = _activeItem.trace;

        var stage = state.map(function (i) {
          return item(i, Object.assign({ id: id, trace: trace }, action));
        });

        // Trace also keeps parentId nonetheless it doesn't matter
        return content(stage, (0, _content.changeSubMenuVisibility)(parentId, trace, true));
      }
    default:
      {
        return state;
      }
  }
};

exports.default = content;