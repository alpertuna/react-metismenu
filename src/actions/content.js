/**
 * @file actions/content.js
 * @author H.Alper Tuna <halpertuna@gmail.com>
 * Date: 16.12.2016
 */

export const updateContent = (reduxUid, content) => ({
  type: 'UPDATE_CONTENT',
  reduxUid,
  content,
});

export const changeSubMenuVisibility = (reduxUid, id, trace, subMenuVisibility) => ({
  type: 'CHANGE_SUBMENU_VISIBILITY',
  reduxUid,
  id,
  trace,
  subMenuVisibility,
});

export const changeActiveLinkId = (reduxUid, value) => ({
  type: 'CHANGE_ACTIVE_LINK',
  reduxUid,
  propName: 'id',
  value,
});
export const changeActiveLinkTo = (reduxUid, value) => ({
  type: 'CHANGE_ACTIVE_LINK',
  reduxUid,
  propName: 'to',
  value,
});
export const changeActiveLinkLabel = (reduxUid, value) => ({
  type: 'CHANGE_ACTIVE_LINK',
  reduxUid,
  propName: 'label',
  value,
});
export const changeActiveLinkFromLocation = reduxUid => ({
  type: 'CHANGE_ACTIVE_LINK_FROM_LOCATION',
  reduxUid,
});

var changeMainMenuContentVisibility = exports.changeMainMenuContentVisibility = function changeMainMenuContentVisibility(reduxUid) {
  return {
    type: 'CHANGE_MAIN_MENU_CONTENT_VISIBILITY',
    reduxUid: reduxUid
  };
};
