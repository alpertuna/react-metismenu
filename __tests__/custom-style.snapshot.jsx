/**
 * custom-style.snapshot.jsx
 * Author: H.Alper Tuna <halpertuna@gmail.com>
 * Date: 15.09.2016
 */

/* eslint-env jest */
/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */

import React from 'react';
import renderer from 'react-test-renderer';
import Menu from '../src';
import content from './CONTENT.json';

describe('Menu', () => {
  it('with no built-in classes', () => {
    const component = renderer.create(<Menu content={content} noBuiltInClassNames />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('with customized class names', () => {
    const customClassNamesProp = {
      className: 'test-c',
      classNameContainer: 'test-cc',
      classNameContainerVisible: 'test-ccv',
      classNameItem: 'test-ci',
      classNameLink: 'test-cl',
      /* classNameLinkActive: 'test-cla',
      classNameLinkHasActiveChild: 'test-clhac', */
      classNameIcon: 'test-cic',
      classNameStateIcon: 'test-csi',
      iconNamePrefix: 'test-ip',
      iconNameStateVisible: 'test-isv',
      iconNameStateHidden: 'test-ish',
    };

    it('and built-in classes', () => {
      const component = renderer.create(<Menu content={content} {...customClassNamesProp} />);
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
    it('and no built-in classes', () => {
      const component = renderer.create(
        <Menu content={content} noBuiltInClassNames {...customClassNamesProp} />
      );
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
    it('classNameContainer func support', () => {
      customClassNamesProp.classNameContainer = itemId => itemId ? 'child-cc' : 'test-cc'
      const component = renderer.create(
        <Menu content={content} noBuiltInClassNames {...customClassNamesProp} />
      );
      const tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
