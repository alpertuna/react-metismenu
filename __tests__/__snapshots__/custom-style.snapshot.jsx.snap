exports[`Menu with customized class names and built-in classes 1`] = `
<div
  className="metismenu test-c">
  <ul
    className="metismenu-container test-cc">
    <li
      className="metismenu-item test-ci">
      <a
        className="metismenu-link test-cl"
        href="#item-1"
        onClick={[Function activateMe]}
        target={undefined}>
        <i
          className="metismenu-icon test-cic test-ipbolt" />
        Item 1
      </a>
    </li>
    <li
      className="metismenu-item test-ci">
      <a
        className="metismenu-link test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="metismenu-icon test-cic test-ipbell" />
        Item 2
        <i
          className="metismenu-state-icon test-csi test-iptest-ish" />
      </a>
      <ul
        className="metismenu-container test-cc">
        <li
          className="metismenu-item test-ci">
          <a
            className="metismenu-link test-cl"
            href="#sub-item-1"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="metismenu-icon test-cic test-ipeye" />
            Sub Item 1
          </a>
        </li>
        <li
          className="metismenu-item test-ci">
          <a
            className="metismenu-link test-cl"
            href="#sub-item-2"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="metismenu-icon test-cic test-ipglobe" />
            Sub Item 2
          </a>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item test-ci">
      <a
        className="metismenu-link test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="metismenu-icon test-cic test-ipbell" />
        Item 3
        <i
          className="metismenu-state-icon test-csi test-iptest-ish" />
      </a>
      <ul
        className="metismenu-container test-cc">
        <li
          className="metismenu-item test-ci">
          <a
            className="metismenu-link test-cl"
            href="#sub-item-3"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="metismenu-icon test-cic test-ipeye" />
            Sub Item 3
          </a>
        </li>
        <li
          className="metismenu-item test-ci">
          <a
            className="metismenu-link test-cl"
            href="#"
            onClick={[Function toggleSubMenu]}
            target={undefined}>
            <i
              className="metismenu-icon test-cic test-ipglobe" />
            Sub Item 4
            <i
              className="metismenu-state-icon test-csi test-iptest-ish" />
          </a>
          <ul
            className="metismenu-container test-cc">
            <li
              className="metismenu-item test-ci">
              <a
                className="metismenu-link test-cl"
                href="#level-3-item-1"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="metismenu-icon test-cic test-ipbolt" />
                Level 3 Item 1
              </a>
            </li>
            <li
              className="metismenu-item test-ci">
              <a
                className="metismenu-link test-cl"
                href="#level-3-item-2"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="metismenu-icon test-cic test-ipbell" />
                Level 3 Item 2
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className="metismenu-item test-ci">
      <a
        className="metismenu-link test-cl"
        href="http://www.google.com"
        onClick={[Function activateMe]}
        target="_blank">
        <i
          className="metismenu-icon test-cic test-ipexternal-url" />
        External Url
      </a>
    </li>
  </ul>
</div>
`;

exports[`Menu with customized class names and no built-in classes 1`] = `
<div
  className="test-c">
  <ul
    className="test-cc">
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#item-1"
        onClick={[Function activateMe]}
        target={undefined}>
        <i
          className="test-cic test-ipbolt" />
        Item 1
      </a>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="test-cic test-ipbell" />
        Item 2
        <i
          className="test-csi test-iptest-ish" />
      </a>
      <ul
        className="test-cc">
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-1"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipeye" />
            Sub Item 1
          </a>
        </li>
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-2"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipglobe" />
            Sub Item 2
          </a>
        </li>
      </ul>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="test-cic test-ipbell" />
        Item 3
        <i
          className="test-csi test-iptest-ish" />
      </a>
      <ul
        className="test-cc">
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-3"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipeye" />
            Sub Item 3
          </a>
        </li>
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#"
            onClick={[Function toggleSubMenu]}
            target={undefined}>
            <i
              className="test-cic test-ipglobe" />
            Sub Item 4
            <i
              className="test-csi test-iptest-ish" />
          </a>
          <ul
            className="test-cc">
            <li
              className="test-ci">
              <a
                className="test-cl"
                href="#level-3-item-1"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="test-cic test-ipbolt" />
                Level 3 Item 1
              </a>
            </li>
            <li
              className="test-ci">
              <a
                className="test-cl"
                href="#level-3-item-2"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="test-cic test-ipbell" />
                Level 3 Item 2
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="http://www.google.com"
        onClick={[Function activateMe]}
        target="_blank">
        <i
          className="test-cic test-ipexternal-url" />
        External Url
      </a>
    </li>
  </ul>
</div>
`;

exports[`Menu with customized class names classNameContainer func support 1`] = `
<div
  className="test-c">
  <ul
    className="test-cc">
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#item-1"
        onClick={[Function activateMe]}
        target={undefined}>
        <i
          className="test-cic test-ipbolt" />
        Item 1
      </a>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="test-cic test-ipbell" />
        Item 2
        <i
          className="test-csi test-iptest-ish" />
      </a>
      <ul
        className="child-cc">
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-1"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipeye" />
            Sub Item 1
          </a>
        </li>
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-2"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipglobe" />
            Sub Item 2
          </a>
        </li>
      </ul>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="test-cic test-ipbell" />
        Item 3
        <i
          className="test-csi test-iptest-ish" />
      </a>
      <ul
        className="child-cc">
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#sub-item-3"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="test-cic test-ipeye" />
            Sub Item 3
          </a>
        </li>
        <li
          className="test-ci">
          <a
            className="test-cl"
            href="#"
            onClick={[Function toggleSubMenu]}
            target={undefined}>
            <i
              className="test-cic test-ipglobe" />
            Sub Item 4
            <i
              className="test-csi test-iptest-ish" />
          </a>
          <ul
            className="child-cc">
            <li
              className="test-ci">
              <a
                className="test-cl"
                href="#level-3-item-1"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="test-cic test-ipbolt" />
                Level 3 Item 1
              </a>
            </li>
            <li
              className="test-ci">
              <a
                className="test-cl"
                href="#level-3-item-2"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="test-cic test-ipbell" />
                Level 3 Item 2
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className="test-ci">
      <a
        className="test-cl"
        href="http://www.google.com"
        onClick={[Function activateMe]}
        target="_blank">
        <i
          className="test-cic test-ipexternal-url" />
        External Url
      </a>
    </li>
  </ul>
</div>
`;

exports[`Menu with no built-in classes 1`] = `
<div
  className="">
  <ul
    className="">
    <li
      className="">
      <a
        className=""
        href="#item-1"
        onClick={[Function activateMe]}
        target={undefined}>
        <i
          className="fa fa-bolt" />
        Item 1
      </a>
    </li>
    <li
      className="">
      <a
        className=""
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="fa fa-bell" />
        Item 2
        <i
          className="fa fa-caret-left" />
      </a>
      <ul
        className="">
        <li
          className="">
          <a
            className=""
            href="#sub-item-1"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="fa fa-eye" />
            Sub Item 1
          </a>
        </li>
        <li
          className="">
          <a
            className=""
            href="#sub-item-2"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="fa fa-globe" />
            Sub Item 2
          </a>
        </li>
      </ul>
    </li>
    <li
      className="">
      <a
        className=""
        href="#"
        onClick={[Function toggleSubMenu]}
        target={undefined}>
        <i
          className="fa fa-bell" />
        Item 3
        <i
          className="fa fa-caret-left" />
      </a>
      <ul
        className="">
        <li
          className="">
          <a
            className=""
            href="#sub-item-3"
            onClick={[Function activateMe]}
            target={undefined}>
            <i
              className="fa fa-eye" />
            Sub Item 3
          </a>
        </li>
        <li
          className="">
          <a
            className=""
            href="#"
            onClick={[Function toggleSubMenu]}
            target={undefined}>
            <i
              className="fa fa-globe" />
            Sub Item 4
            <i
              className="fa fa-caret-left" />
          </a>
          <ul
            className="">
            <li
              className="">
              <a
                className=""
                href="#level-3-item-1"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="fa fa-bolt" />
                Level 3 Item 1
              </a>
            </li>
            <li
              className="">
              <a
                className=""
                href="#level-3-item-2"
                onClick={[Function activateMe]}
                target={undefined}>
                <i
                  className="fa fa-bell" />
                Level 3 Item 2
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className="">
      <a
        className=""
        href="http://www.google.com"
        onClick={[Function activateMe]}
        target="_blank">
        <i
          className="fa fa-external-url" />
        External Url
      </a>
    </li>
  </ul>
</div>
`;
